src/protocol/mod.rs

```diff
@@ -222,11 +222,22 @@ impl Protocol {
         })
     }
 
-    pub fn load(executable_path: &Path) -> R<Vec<Protocol>> {
-        let file_contents = read_protocols_file(executable_path)?;
+    fn parse_protocols(file_contents: &str) -> R<Vec<Protocol>> {
         let yaml: Vec<Yaml> = YamlLoader::load_from_str(&file_contents)?;
         yaml.into_iter().map(Protocol::parse).collect()
     }
+
+    pub fn load(executable_path: &Path) -> R<Vec<Protocol>> {
+        let file_contents = read_protocols_file(executable_path)?;
+        let result = Protocol::parse_protocols(&file_contents).map_err(|error| {
```

Comment:
is it necessary to unwrap the result with `?`, assign it, and then rewrap with `Ok(result)` below?

Reply:
Yeah, it is. `map_err` returns a value of type `Result<Vec<Protocol>, String>`. The result of `load` is `R<Vec<Protocol>>` which expands to `Result<Vec<Protocol>, Box<std::error::Error>>`. The `?` implicitly converts the error with the `From` trait. So we need that.

What would work is this:

``` rust
Ok(Protocol::parse_protocols(&file_contents).map_err(|error| {
    format!(
        "error parsing {}: {}",
        add_extension(executable_path, "protocols.yaml").to_string_lossy(),
        error
    )
})?)
```

Would that be better?

Reply:
ah, implicits. hmm, i don't think that would be any better, i think this is always bound to happen when implicit stuff occurs (where some lines of code aren't immediately obvious). in my mind the ideal thing to see would be a more explicit conversion, like...

```rust
Protocol::parse_protocols(&file_contents).map_err(|error| {
    format!(
        "error parsing {}: {}",
        add_extension(executable_path, "protocols.yaml").to_string_lossy(),
        error
    ).to_boxed() // here
})
```
but that could just be my lack of experience with rust